core/base/inc/TPluginManager.h

```diff
@@ -146,18 +150,14 @@ friend class TPluginManager;
 
    template <typename... T> Longptr_t ExecPluginImpl(const T&... params)
    {
-      auto nargs = sizeof...(params);
+      constexpr auto nargs = sizeof...(params);
       if (!CheckForExecPlugin((Int_t)nargs)) return 0;
 
-      // The fCallEnv object is shared, since the PluginHandler is a global
-      // resource ... and both SetParams and Execute ends up taking the lock
-      // individually anyway ...
-
-      R__LOCKGUARD(gInterpreterMutex);
-      fCallEnv->SetParams(params...);
-
       Longptr_t ret;
-      fCallEnv->Execute(ret);
+      const void *args[nargs] = {&params...};
```

Comment:
Yes that would explain the test failures.  Let me find some middle ground here...

Reply:
I was probably too terse :(.  It is not reintroducing literally the old behavior but in only a similar fashion.   In this new code, the call will work only in the case where the arguments are const char* or const char* containing text to be parsed to get the real values.  (i.e. matching exactly the argument type doesn't help the new call).  Sorry about the confusion.     

However, if we know no conversion needs to be used, we should be able to use a short cut and use directly the function pointer returned by `TMethodCall::GetCallFunc`.

For the other cases, since `TClingCallFunc` caches the `cling::Value` corresponding to the actual input (and later applying the conversion as needed), the easiest is likely to create a copy of the `TMethodCall` (on the stack?) and use it instead of the 'global' one. 

And this until a stateless code path is added (i.e we could merge the `SetParams` and `Execute` in TMethodCall and thus `CallFunc_SetArguments` and `Execute` in TInterpreter so that they act on a "local"(ish?) `llvm::SmallVector<cling::Value, 8> ` rather than the `TClingCallFunc`

Note: the copying of `TMethodCall` is not 'light'.

Reply:
Yah the problem is that there also plenty of global locks within ```TClingCallFunc``` while doing the conversion etc.  Indeed the "alternative" solution I've implemented for the moment is along the lines of what you suggest with a fast path when no conversion is needed (the verification of this is a bit kludgy though)